### PR TITLE
Jumpstarter

### PR DESCRIPTION
**Pull request by MicahLyle**
At the time of writing (2021-05-09), very WIP initial draft. I only filled out an abstract which I think is probably too long, but I wanted to get my thoughts out there and hopefully at least start discussing things.

The motivation I also started to fill out. I'm not entirely sure where I want to go with that exactly yet, but wanted to get things going. I think it's important to me to show how the actor model and actor system can be a really great foundation for next-gen Celery, and also clearly explain why it's a win vs. other trade-offs.

I'm not familiar enough, at this point, with the internal details of a lot of the Celery components enough, in my opinion, to make authoritative claims about certain things, etc., but I do really think that composable tasks and messages via an actor system is the way to go, and I want to argue that case well, etc.

### CODE REVIEW

File: draft/jumpstarter.rst [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+Abstract
+========
+
+`Next-Gen Celery`_ aims to:
+
+1. Move Celery from the past to the present. 
+
+2. Resolve long-standing design bugs in our implementation. 
+
+3. Modernize the code to support Python3+.
```
Comment (auvipy) [inline draft/jumpstarter.rst:29]: python 3.6/3.7+?

**Review by auvipy — COMMENTED** [review]
good start

File: draft/jumpstarter.rst (hunk 1 above)
Comment (thedrow) [reply draft/jumpstarter.rst:29]: We don't know yet but it's likely to be 3.7+ or even 3.10+ (in case we want to use pattern matching).

Comment (MicahLyle) [reply draft/jumpstarter.rst:29]: 3.10+ may also have some niceities with async libraries although I don't think exception groups is landing in 3.10 (https://www.python.org/dev/peps/pep-0654/)

Comment (thedrow) [reply draft/jumpstarter.rst:29]: That will probably hit 3.11 or higher and it will take time for Trio and AnyIO to adapt.

File: draft/jumpstarter.rst [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+1. How does Celery interact with itself (its own internal components)? 
+
+2. How does Celery interact with others (and how do others interact with it)? 
+
+3. How do Celery's internal components interact with each other and their outside environment? 
+
+Also, we want Celery to work with the emerging and maturing asynchronous python
+landscape.  Celery is *very asynchronous* by nature, dealing with brokers,
+queues, results, timeouts, chord-like joins, task groups, you name it.
+Modern asynchronous python frameworks have provided new, innovative, and robust
```
Comment (thedrow) [inline draft/jumpstarter.rst:53]:
```suggestion
Modern asynchronous Python frameworks have provided new, innovative, and robust
```

File: draft/jumpstarter.rst [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+solutions to many things that overlap with Celery's goals.
+
+To do all of this, we propose modeling next-gen Celery off of the `Actor Model`_. Namely,
+we propose that the Celery Worker will be modeled as an `Actor System`_ of sorts.
+
+By the end of this initial implementation, we should have a proposal and reference implementation that: 
+
+1. Models Tasks and other key Celery primitives (some to be implemented and
```
Comment (thedrow) [inline draft/jumpstarter.rst:61]: I think we should reduce scope to first get a working Actor framework with all the features it needs to function as a standalone framework and then starts modeling Celery primitives as actors.

File: draft/jumpstarter.rst [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+further specced-out down the line) as Actors. 
+
+2. Can give a clear overview about how these various Celery primitives (as
+Actors) communicate with each other and pass messages (referencing the Actor
+model). 
+
+3. Can enable the development of a future version of Celery where it can be run
```
Comment (thedrow) [inline draft/jumpstarter.rst:68]: That should be goal 1 in my opinion.

File: draft/jumpstarter.rst [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+Motivation
+==========
+
+TODO (WIP)
+
+This section should explain *why* this CEP is needed. The motivation is critical
+for CEPs that want to add substantial new features or materially refactor
+existing ones.  It should clearly explain why the existing solutions are
+inadequate to address the problem that the CEP solves.  CEP submissions without
+sufficient motivation may be rejected outright.
```
Comment (thedrow) [inline draft/jumpstarter.rst:94]:
```suggestion
```

File: draft/jumpstarter.rst [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+Let's start with an example, and then generalize. A quick search in Celery's
+GitHub issue tracker for `chord
+<https://github.com/celery/celery/issues?q=is%3Aissue+is%3Aopen+chord>`_ shows
+a many issues. A number of them (of which some have been fixed) involve complex
+cases involving non-trivial Celery workflows. While fixes have been released
+over time as these have come up, many issues stand unresolved to this day. The
+Celery team could continue to try and fix these issues, or we could take a look
+back and think about how we want to design Celery for the future. although
```
Comment (thedrow) [inline draft/jumpstarter.rst:103]:
```suggestion
back and think about how we want to design Celery for the future. Although
```

File: draft/jumpstarter.rst [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+there have been a number of difficulties with getting started on fixing these
+issues in the first place! I'll try and list some of what I see these to be:
```
Comment (thedrow) [inline draft/jumpstarter.rst:105]: This sentence should probably be rephrased to a more formal language.

File: draft/jumpstarter.rst [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+2. Celery, as a whole, involves a number of different working parts that are
+integrated together. In the current scheme of things, we have workers, brokers,
+result stores, workflows (``canvas`` primitives), rate limiting, policies,
+signatures, settings, periodic tasks, scheduling, ..., and many other things.
+Many of the primitives for these integrated parts were built almost a decade
+ago, when both the language landscape (as mentioned above), but also the
+*library landscape* was quite a bit different than today. Celery came with a
+custom CLI parser, custom event loop implementation (which it still has at the
+time of writing), and support for a number of other things that are today
+implemented in multiple different solid third party libraries.
```
Comment (thedrow) [inline draft/jumpstarter.rst:125]: 100% correct. Well done.

File: draft/jumpstarter.rst [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+root cause of what's happening in the first place. Because Celery is very
+*asynchronous* by nature, many of the tests are wrapped with a ``flaky``
+decorator which means, sometimes (depending on the arrival of messages, speed
+of IPC, speed of the network, and many other factors), maybe the test won't
+pass in an expected amount of time or might not always succeed. Testing and
+modeling asynchronous processes is definitely not easy, but the python
+language, tooling, and libraries have evolved considerably in the last
+*decade*, and Celery is well positioned for a large restructuring that allows
+it both internally take advantage of the latest Python features, and also
+provide external integration with them as well.
```
Comment (thedrow) [inline draft/jumpstarter.rst:137]: Perfect. 2 & 3 are exactly what I had in mind.

File: draft/jumpstarter.rst [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+number of benefits, which I want to succintly summarize in the beginning and
+give a lot more detail as a part of the CEP. Specifically, I'd like to
+explain/say why modeling things as an Actor system will make Celery a lot
+easier to reason about, maintain, and add new, for example, Canvas primitives
+to (or even re-build/design them, etc. if that's something that's desired).
+There are clear advantages to keeping state internal to the worker, for
+example, adding cancel groups, and other things that make some of the more
+asynchronous parts of Celery easier to work with, test, and reason about. The
+main thing also, though, is to really separate concerns and responsibilities.
+It *seems* to me that some of the current Celery code is not very SRP, in the
```
Comment (thedrow) [inline draft/jumpstarter.rst:157]: > It *seems* to me that some of the current Celery code is not very SRP

That's one of the main problems of the codebase.

File: draft/jumpstarter.rst [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+and doing a number of complex things. With the actor system, each actor gets a
+lot more specialized and I think that makes the implementation a lot easier to
+both reason about and extend.  Composition (vs., for example, inheritance
+currently present with ``Signature`` objects for example to then create the
+canvas objects) then becomes a lot more attractive and possible.  To give an
+example, a ``chord`` then simply becomes responsible for passing a message to
+the underlying ``Task`` (s) or ``result`` (s) (to define more) that would
+essentially make them then send off another "message" (we'll call it right now)
+or messages to other components in the actor system that are then responsible
+for handling that message (think of ``on_chord_part_return`` here).
```
Comment (thedrow) [inline draft/jumpstarter.rst:168]: Agreed.

File: draft/jumpstarter.rst (hunk 3 above)
Comment (MicahLyle) [reply draft/jumpstarter.rst:61]: Agreed. Working on that next. The next draft should have that.

File: draft/jumpstarter.rst (hunk 1 above)
Comment (MicahLyle) [reply draft/jumpstarter.rst]: Another nice thing about `3.10` in my opinion is that it makes:

``from __future__ import annotation`` completely optional, and 3.10 has a number of type-hinting improvements.

How important do you think being able to have a nicely type-hinted codebase is, vs. supporting 3.7/3.8/3.9+? We could do the `typing_extensions` workaround. But I've really loved seeing @thedrow's jumpstarter code with the very modern type hints, including `|` usage, self-referential annotations, etc. I feel like 3.10 is a pretty big milestone with that.

All of that being said though, this can all be emulated with the `__future__` import and `typing_extensions` I think.

File: draft/jumpstarter.rst (hunk 3 above)
Comment (MicahLyle) [reply draft/jumpstarter.rst:61]: I think the trickiest part at the moment for me is trying to figure out how much to talk about Jumpstarter and how much to talk about Celery. Do you feel like this CEP should be more focused on the whys and what's of a working Actor framework in Python, or the whys and whats of a future async Celery and the rationale for building an underlying standalone framework that would "power" Celery (but also contribute something to the broader Python community in terms of a missing actor framework).?

File: draft/jumpstarter-prev-draft.rst [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+
+TODO (WIP)
+
+This section should explain *why* this CEP is needed. The motivation is critical
+for CEPs that want to add substantial new features or materially refactor
+existing ones.  It should clearly explain why the existing solutions are
+inadequate to address the problem that the CEP solves.  CEP submissions without
+sufficient motivation may be rejected outright.
+
+Let's start with an example, and then generalize. A quick search in Celery's
```
Comment (MicahLyle) [inline draft/jumpstarter-prev-draft.rst:96]: @thedrow I'd like to potentially move this section back into the CEP probably in the next revision. However, I wanted to ask if you think this is a good idea or not. Is it too out of scope? I really like a bit of the discussion on Celery's internals and issues that could be resolved by more SRP code, and why something new is needed, but maybe it's out of scope and should be reserved for a future CEP?

**Comment by MicahLyle** [comment]
Second draft. I have not filled in the specification yet. That I'm planning to do next, and getting feedback on the higher-level direction of the CEP has been really helpful so far. This one aimed to reduce the scope a bit, and I temporarily have stored the previous draft in a different document for comparison and so that I could pull things back in as needed. I wanted to see a more simplified CEP, both in response to @thedrow's suggestion, but also to see how it would feel overall.

I'm really happy about the Rationale section. I feel like it was important to review the current landscape and give clear reasons as to why we'd roll our own library than use something else out there.

File: draft/jumpstarter-prev-draft.rst (hunk 12 above)
Comment (MicahLyle) [reply draft/jumpstarter-prev-draft.rst:96]: Also, if I did, I would rephrase things in more formal language in certain sections per your comment.

**Comment by MicahLyle** [comment]
One other thought. There may be sections from the previous draft to merge back into the current draft and/or future drafts. However, I wanted to remove them for this second draft to see what it might look like with a trimmed down scope. I liked @thedrow's comment about reducing to scope to `async/await` and supporting a future version of Celery that could interact with `async/await` on a number of levels, so this draft explores a reduced scope in that regard. As mentioned in the comments on the changed files from a little bit ago, there are some sections I think I'd like to see included somewhere, whether this CEP or a future CEP.

File: draft/jumpstarter-prev-draft.rst [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+library support. One example is we're waiting for an `APScheduler 4.0
+Release`_.
+
+Copyright
+=========
+
+This document has been placed in the public domain per the Creative Commons
+CC0 1.0 Universal license (https://creativecommons.org/publicdomain/zero/1.0/deed).
+
+(All CEPs must include this exact copyright statement.)
```
Comment (thedrow) [inline draft/jumpstarter-prev-draft.rst:210]: FYI, this is part of the template.
```suggestion
```

File: draft/jumpstarter-prev-draft.rst [hunk 14]
```diff
@@ -0,0 +1,219 @@
+.. vale off
+
```
Comment (thedrow) [inline draft/jumpstarter-prev-draft.rst:2]:
```suggestion
```

File: draft/jumpstarter.rst [hunk 15]
```diff
@@ -20,175 +20,143 @@ CEP XXXX: Jumpstarter
...
+
+1. The motivation to build `Jumpstarter`_.
+
+2. The motivation to, down the line, use `Jumpstarter`_ as a foundation for parts of
+`Next-Gen Celery`_.
+
+For the first motivation, one of Celery's main use cases is to build asynchronous,
+distributed systems that communicate via message passing. The `Actor Model`_, which has
+been around for almost half a century and is a tried and tested way to design and build
+large-scale concurrent systems. It very much matches what Celery aims to do and has
```
Comment (thedrow) [inline draft/jumpstarter.rst:82]:
```suggestion
distributed systems that communicate via message passing. The `Actor Model`_, which has
been around for almost half a century is a tried and tested way to design and build
large-scale concurrent systems. It very much matches what Celery aims to do and has
```

File: draft/jumpstarter.rst [hunk 16]
```diff
@@ -20,175 +20,143 @@ CEP XXXX: Jumpstarter
...
-the underlying ``Task`` (s) or ``result`` (s) (to define more) that would
-essentially make them then send off another "message" (we'll call it right now)
-or messages to other components in the actor system that are then responsible
-for handling that message (think of ``on_chord_part_return`` here).
+  * It doesn't support ``async/await`` currently, and hasn't supported it from the beginning.
+  * It powers `Mopidy`_, and probably a number of other significant projects rely on it to some extent, so it wouldn't make sense to rely upon it for reasons listed above.
 
+* `Cell`_ was an earlier attempt at an actor model/framework for Celery. It wasn't very widely used and developed.
 
+  * Given reason #1 above, it makes sense to archive `Cell`_ and move forward with `Jumpstarter`_ (`comment <https://github.com/celery/jumpstarter/issues/1#issuecomment-755347761>`_).
```
Comment (thedrow) [inline draft/jumpstarter.rst]: Note that Cell implemented actors **on top** of the existing Celery implementation.
You couldn't use it to implement Celery itself.

**Review by thedrow — COMMENTED** [review]
💯

File: draft/jumpstarter.rst [hunk 17]
```diff
@@ -64,6 +64,115 @@ the other features of Python's ``async/await`` programming model.
 Specification
 =============
 
+Jumpstarter is a Python implementation of an `Actor System`_ (which utilizies the `Actor Model`_). There
+are three fundamental axioms within the actor model (quoting the previous Wikipedia link): 
+
+(Start Blockquote)
```
Comment (thedrow) [inline draft/jumpstarter.rst:70]: See [RST spec](https://docutils.sourceforge.io/docs/user/rst/quickref.html#block-quotes) for blockqoutes.

File: draft/jumpstarter.rst [hunk 18]
```diff
@@ -64,6 +64,115 @@ the other features of Python's ``async/await`` programming model.
...
+
+(End Blockquote)
+
+It's important to remember that, although that is the technical definition of the actor, the interpretation and implementation of Actors and Actor Systems can be very flexible. Namely, what constitutes a "message" and "state" is very much up to the interpretation of the developer(s) and the system(s) they're using.
+
+In Jumpstarter, we've chosen to take direct/literal approach to 3., modeling the state of an Actor using an actual state machine abstraction, namely a `Hierarchical State Machine`_. The difference between a standard State Machine and a Hierarchical State Machine is that a standard State Machine is consistent from states and transitions between them, but in an Hierarchical State Machine, states can also have their own sub-state machines. Hierarchical State Machines help both tame the complexity of large (non-hierarchical) state machines and more clearly model the relationships and transitions between them. To give an example with Jumpstarter, we propose only a small number of parent states:
+
+* Initializing --> The initial state of the Actor when created.
+* Initialized --> he state of the actor when we start it for the very first time.
```
Comment (thedrow) [inline draft/jumpstarter.rst:85]:
```suggestion
* Initialized --> The state of the actor when we start it for the very first time.
```

File: draft/jumpstarter.rst [hunk 19]
```diff
@@ -64,6 +64,115 @@ the other features of Python's ``async/await`` programming model.
...
+3. Native support within `transitions`_ for integrating with ``diagrams``/``graphviz`` to generate state machine diagrams (like the one below). Additionally, `transitions-gui`_ provides some interesting and promising capabilities for future Celery Flower-like projects to be able to visualize in a live, animated fashion the various Jumpstarter Actors and their states as transitions happen across all the various actors within the system.
+
+For a high level view, the parent states, their substates, and the transitions between them can be seen in the diagram below:
+
+TODO: Insert Jumpstarter State Machine Diagram Here: https://user-images.githubusercontent.com/48936/107506089-43225a00-6ba6-11eb-810e-0ac14bf0e1e9.png
+
+Also, in that diagram you can also see the ``Restart`` state. We propose a separate state machine which we'll call *Actor Restart State Machine* that models the Actor's state as it relates to restarts:
```
Comment (thedrow) [inline draft/jumpstarter.rst:118]: This should be a parallel state in our state machine but transitions doesn't support arbitrary parallel states, only initial parallel states.
This is an open item in our implementation to resolve.
Currently, the draft implementation nests state machines in the parent state machine and we model transitions using conditions on the transition itself.

File: draft/jumpstarter.rst [hunk 20]
```diff
@@ -64,6 +64,115 @@ the other features of Python's ``async/await`` programming model.
...
+class AccountBookkeepingActor(Actor):
+  def __init__(self, user_id: int, account_balance_actor: AccountBalanceActor):
+    self._account_balance_actor = account_balance_actor
+
+  @depends_on
+  def account_balance_actor(self):
+    return account_balance_actor
+```
+
+In this example, the ``AccountBalanceActor`` maintains the balance in a single user ID's account. The ``AccountBookkeepingActor`` is responsible for logging and auditing withdrawals and income, possibly passing these audit logs to another actor responsible for detecting fraud.
```
Comment (thedrow) [inline draft/jumpstarter.rst:162]: This is still up for discussion. I haven't really given much thought to dependencies.

**Review by thedrow — CHANGES_REQUESTED** [review]
Glad to see some progress here.

**Comment by MicahLyle** [comment]
@thedrow at this point, I'd say that I have an "initial draft", complete from the perspective of taking the material you had across the Jumpstarter issues (and code), trying to scope it a bit, and getting something out there. I haven't tried to come up with anything on my own at this point (minus the example at the end of the specification), but rather just tried to solidify and summarize what you've already done.

The Specification feels bare/basic and mostly copy pasted (with some edits and changes as I felt would be helpful) from the various work you had around the issues. I know they were rough guidelines, but there are certain things I don't fully understand, so it was hard to write about them besides just trying to put down what you already had and clarify things a bit.

Now that I've had plenty of time to really dig into your specifications, dig into the `transitions` library, and do some other research on actors, etc., here are some unresolved questions I have and general thoughts:

- I don't understand `continuous` tasks per-say. It's a little hard to reason about the idea of a task just repeatedly executing until the state is executed, _unless_, I'm misunderstanding something. In your producers/consumers spec, does returning `Success` indicate to some parent actor/executor that the task should be complete, causing it to trigger, say a `stop()` transition? Meaning, if you don't wrap in `Success`, you're effectively just "producing" values (continuously) until you see `Success`?
- I feel like using `transitions` is a very novel approach, and I like the idea of modeling everything after state and transitions (after all, Hewitt and others mention actors as just having simple local state usually, and if everything has state, using an actual state machine abstraction seems to fit really nicely). _That being said_, grokking the `transitions` code, subtleties, and other things was not straightforward or trivial. Additionally, even grokking the Jumpstarter code and tests, as barebones and simple as they are, took a bit of time in terms of wrapping my head around what context managers were being entered (and exited when), which transitions were triggering which things, etc. My concern here is that if/when a public API is introduced, I think, unless someone wants to heavily customize it, it would be important to keep things simple and provide some guarantees.
- I think `transitions` `queued=True` mode makes sense, as you've used in in the draft. _That being said_, the way that it hides exceptions that can occur, especially in a system where we're relying on context managers to manage certain sections of code and be entered and exited at proper times, makes me a little nervous. Additionally, I feel like some of the API and boundaries of `transitions`, especially when it gets to the hierarchical state machines and parallel state machines, are rough around the edges and could lead to some hard to debug edge cases and error handling, especially around timeouts, proper context manager cleanup and resource handling, and other things like that. The API is still at a state where I feel like it is hard to understand and use properly, and I would like Jumpstarter to be something that is easy to get started with (like `trio` for example took a complex thing like `async/await` and made it very simple to use and understand, while also exposing more complex APIs that you could reach for when needed). I like the extension flexibility, but the complexity does seem quite high at the moment.
- Do you think that just having resources, dependencies, and tasks initially defined with a relatively barebones and simple explanations and examples is enough for this CEP for now? Or would you want producers and consumers specced out as well? I felt like the producers and consumers section of the GitHub Issues on Jumpstarter felt the least clear, and in trying to just have a base of what was already written, I felt hesitant to include it in this CEP because it seems that it might be good to work out the API as an implementation detail (and just describe a short, high level abstract of what we'd like to have the capability to do) rather than trying to figure out how we're going to get it to work in the CEP.

I may add some other questions and comments if I think of more, but I'd like to have a discussion around these things at least.

File: draft/jumpstarter.rst [hunk 21]
```diff
@@ -143,24 +143,30 @@ Actors may depend on other actors to run before starting themselves. In some cas
...
+      # an `AccountBalanceActor`.
+      return account_balance_actor
 ```
 
 In this example, the ``AccountBalanceActor`` maintains the balance in a single user ID's account. The ``AccountBookkeepingActor`` is responsible for logging and auditing withdrawals and income, possibly passing these audit logs to another actor responsible for detecting fraud.
 
+Instead of returning an already existing *instance* of an ``AccountBalanceActor`` in ``@depends_on``, you can also:
+1. Use a factory method to initialize a brand new ``AccountBalanceActor`` instance (since every actor must inherit from ``Actor`` we'll define some helpful factory methods in ``Actor`` which can be used by all subclasses/instances).
+2. Return a subclass of ``Actor`` and it will be initialized for you, proiding all the arguments are available for that actor. This uses the `Inversion of Control`_ pattern. How this works will be left as an implementation detail, but Jumpstarter, given that it knows each ``Actor``'s dependencies and has them all in a graph should be able to satisfy dependencies and inject arguments as long as it's able to find them in an accessible way.
```
Comment (auvipy) [inline draft/jumpstarter.rst:212]: proiding or providing?

File: draft/jumpstarter.rst (hunk 1 above)
Comment (auvipy) [reply draft/jumpstarter.rst]: this is a futuristic framework so i dont see problem using 3.10+ as problem. i would rather suggest going 3.10 straight O:) .
